### PR TITLE
Refactor route matching

### LINKED ISSUE #1661: isAmp and service undefined when making client-side requests
**Is your feature request related to a problem? Please describe.**
When fetching data from clicking on an inline like the main data [object](https://github.com/bbc/simorgh/blob/latest/src/app/routes/getInitialData/article.js#L18) is set to `null`. This means that the `isAmp` and the `service` values are not set which are required for setting the request and the service contexts.

So when abstracting the page furniture to its own container the service and isAmp values are not available when in the loading and error states, thus the brand is incorrectly set and isAmp is undefined.

Blocking https://github.com/bbc/simorgh/pull/1589

**Describe the solution you'd like**
We should move the route matching logic which sets the service and isAmp values up into the main App container.

This will enable us to pull the isAmp and the service values out of the data object which is set to null when fetching data.

This change will also remove what would be duplicated logic when setting the service and isAmp values in the getInitial data methods for the article and homepage containers.

**Before**
```
{
      data: this.state.data,
      loading: this.state.loading,
      error: this.state.error,
      bbcOrigin: this.props.bbcOrigin,
}
```

**After**
```
{
      data: this.state.data,
      service: this.state.service,
      isAmp: this.state.isAmp,
      loading: this.state.loading,
      error: this.state.error,
      bbcOrigin: this.props.bbcOrigin,
}
```

**Describe alternatives you've considered**
We could hard-code default values but this would mean that e.g. users may see the wrong error page for their locale, so this is not ideal.

**Testing notes**
This will just require regression of requesting a canonical & an AMP article and checking that there are no errors within the console.

### PR DESCRIPTION
**Pull request by chris-hinds**
Resolves #1661 

**Overall change:**

This PR removes `service` and `isAmp` from the main data object and pulls them into the root state. This way when settings `data` to null during data fetching the application still has access to the `service` and `isAmp` properties.

**Code changes:**

- Pulled the route match logic out of the `loadInitialData` method
- Get the service and isAmp values in the server to pass to the server side rendered document
- get the service and isAmp values on an onward journey during the componentDidUpdate lifecycle method passing the values as state to the main article container.

---

- [ ] I have assigned myself to this PR and the corresponding issues
- [ ] Tests added for new features
- [ ] Test engineer approval
- [ ] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

### CODE REVIEW

File: src/app/containers/App/App.jsx [hunk 1]
```diff
@@ -6,22 +6,35 @@
...
 import { withRouter } from 'react-router-dom';
+import getRouteProps from '../../routes/getInitialData/utils/getRouteProps';
 
 import loadInitialData from '../../routes/loadInitialData';
 
 export class App extends Component {
   constructor(props) {
     super(props);
 
+    const { service, isAmp } = getRouteProps(
```
Comment (chris-hinds) [inline src/app/containers/App/App.jsx:17]: This method is called twice in this class. I have written a new issue suggesting we refactor this file in another PR.

The reason for carrying out this refactor in another PR is that this one is already affecting a lot of files and changes up out main data object. carry out a full refactor of the App class now will add unnecessary complication to this PR.

Issue to further refactor this file: https://github.com/bbc/simorgh/issues/1676

File: src/app/routes/loadInitialData/index.js [hunk 2]
```diff
@@ -2,16 +2,10 @@
  * © Jordan Tart https://github.com/jtart
  * https://github.com/jtart/react-universal-app
  */
-import { matchRoutes } from 'react-router-config';
+import getRouteProps from '../getInitialData/utils/getRouteProps';
 
 const loadInitialData = async (url, routes) => {
```
Comment (chris-hinds) [inline src/app/routes/loadInitialData/index.js:7]: I have proposed to completely remove this method and move the remaining logic into the main App class as a part of another refactor: https://github.com/bbc/simorgh/issues/1676

File: src/server/Document/component.jsx [hunk 3]
```diff
@@ -14,14 +14,20 @@ const Document = ({
...
+  service,
   isAmp,
 }) => {
   const htmlAttrs = helmet.htmlAttributes.toComponent();
   const meta = helmet.meta.toComponent();
   const title = helmet.title.toComponent();
   const links = helmet.link.toComponent();
   const headScript = helmet.script.toComponent();
-  const serialisedData = JSON.stringify(data);
+  const clientDataObj = {
```
Comment (chris-hinds) [inline src/server/Document/component.jsx:25]: We should not need to do this. the client.js files currently uses the service prop on this global object. We should refactor client.js to no longer require this. https://github.com/bbc/simorgh/issues/1679

File: src/app/containers/App/App.test.jsx [hunk 4]
```diff
@@ -11,11 +11,21 @@ describe('App', () => {
...
+  };
+
+  reactRouterConfig.matchRoutes = jest.fn().mockReturnValue([{ route, match }]);
 
   reactRouterConfig.renderRoutes = jest
     .fn()
-    .mockReturnValue(<h1>{initialData.data}</h1>);
+    .mockReturnValue(<h1>{initialData.pageData}</h1>);
+
+  // const getRouteProps = jest.fn().mockReturnValue(match);
```
Comment (rossgaskell) [inline src/app/containers/App/App.test.jsx]: Really tiny, is this still needed or needed in a future refactor?

**Review by rossgaskell — DISMISSED** [review]
Really amazing!

File: src/app/routes/getInitialData/article.js [hunk 5]
```diff
@@ -3,21 +3,18 @@ import getBaseUrl from './utils/getBaseUrl';
...
+  const { id, service } = match.params;
 
   const baseUrl = onClient()
     ? getBaseUrl(window.location.origin)
     : process.env.SIMORGH_BASE_URL;
 
   const url = `${baseUrl}/${service}/articles/${id}.json`;
 
-  const { data, status } = await fetchData({ url });
+  const { pageData, status } = await fetchData({ url });
```
Comment (dr3) [inline src/app/routes/getInitialData/article.js]: Were expanding this object and then rebuilding it below, we can just return directly :)

File: src/app/containers/App/App.test.jsx (hunk 4 above)
Comment (chris-hinds) [reply src/app/containers/App/App.test.jsx]: oops, No this can go. Thanks.

File: src/app/routes/getInitialData/article.js (hunk 5 above)
Comment (chris-hinds) [reply src/app/routes/getInitialData/article.js]: Good spot.

File: src/app/containers/App/App.jsx [hunk 6]
```diff
@@ -6,22 +6,35 @@
...
+      isAmp: isAmp,
       loading: false,
       error: null,
       loadInitialDataPromise: null,
-    };  
+    };
   }
 
   async componentDidUpdate({ location: prevLocation }) {
+    const { service, isAmp } = getRouteProps(
```
Comment (dr3) [inline src/app/containers/App/App.jsx:33]: Not the keenest of this, as any time the component updates we run `getRouteProps` regardless of if we need to or not :/ Not sure of a nice alternative though

Comment (chris-hinds) [reply src/app/containers/App/App.jsx:33]: This could be changed in a new refactor of this whole file, which there is already an issue for.

**Review by dr3 — APPROVED** [review]
lgtm

**Comment by dr3** [comment]
👋 Apologies, this PR needs a bump and changelog update :) Details can be found [here](https://github.com/bbc/simorgh/blob/a75ad1fbf4a720bf6147d7f57ee3d8e3a7337632/CONTRIBUTING.md#versioning-and-changelog)

**Review by jamesdonoh — APPROVED** [review]
Nice job Chris. btw what do you think about us trying to improve the naming of some of these modules/functions when we pick up https://github.com/bbc/simorgh/issues/1679? They don't seem very intuitive to me.

```
loadInitialData
getInitialData
fetchData
```

**Comment by chris-hinds** [comment]
> #1679

@jamesdonoh agree, will add this to the new refactor issue.

**Comment by dr3** [comment]
Was this tested?

**Comment by chris-hinds** [comment]
> Was this tested?

Hi Drew, yes this was tested by @PriyaKR and confirmed that it could be merged.

**Comment by PriyaKR** [comment]
Looks good to me.

**Pull request merged**